src/main/java/com/google/devtools/build/lib/analysis/Runfiles.java

```diff
@@ -1155,65 +1182,32 @@ public Runfiles mergeAll(Sequence<?> sequence, StarlarkThread thread) throws Eva
     }
   }
 
-  /**
-   * Fingerprint this {@link Runfiles} tree.
-   */
-  public void fingerprint(Fingerprint fp) {
+  /** Fingerprint this {@link Runfiles} tree. */
+  public void fingerprint(ActionKeyContext actionKeyContext, Fingerprint fp)
```

Comment:
Mind wrapping this with `catch (InterruptedException) { throw new IllegalStateException(); }`?

My understanding is that the `InterruptedException` part is only needed because `MapFn` can theoretically throw it (but the ones you pass in don't) and I'd rather fail hard if this expectation is violated than to turn an `InterruptedException` into another exception, thereby either swallowing a Ctrl-C or immortalizing it in some cache.

Reply:
A refinement on @lberki 's suggestion would be:
* Declare a new interface that extends `MapFn` but does not have any checked exceptions, say `ExceptionlessMapFn`.
* Have your new `SYMLINK_ENTRY_MAP_FN` and `RUNFILES_AND_EXEC_PATH_MAP_FN` implement this interface.
* Factor out a static helper method that takes in a `ActionKeyContext` and a `ExceptionlessMapFn` and calls `#addNestedSetToFingerprint` and catches the impossible exceptions as lberki@ suggested.
* This way we isolate the assumption to that helper method.

You could also go all the way and add the `ExceptionlessMapFn` outside this file, i.e. add proper support in `ActionKeyContext` and `NestedSetFingerprintCache`.

Reply:
What @haxorz recommended is even better!